Update Tracer#in_span to finish the Span

elskwid opened the pull request:
# Overview

Updates `Tracer#in_span` to call `Span#finish` at the end of the block and adds API tests for `Tracer`.

## Details

While creating some examples based on the DataDog auto-instrumentation work I noticed that the `Tracer#in_span` implementation wasn't finishing the span as expected. I looked into the [Python implementation](https://github.com/open-telemetry/opentelemetry-python/blob/master/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L386) and saw the call the `end()`. This adds the call to `#finish` and adds tests to both the API and SDK.

The tests are going through some slight acrobatics to verify things are working as expected. Please let me know if there are better ways to do this given the implementation(s).

I added tests for the existing API `Tracer` implementation before updating `#in_span` to make sure the changes wouldn't cause us trouble in other methods.

## Thoughts

* I like that the Python API makes the call to `finish`/`end` optional. If we're interested in that I can open a new PR for that functionality. @fbogsany mentioned we may want to do this.
* It's usually bad form to add methods just to ease testing but you can see that I had to poke at the internals a bit to figure out if a span has ended. Any reason not to add `Span#ended?`?

fbogsany commented on api/test/opentelemetry/trace/tracer_test.rb:
```diff
@@ -0,0 +1,118 @@
+      OpenTelemetry::Context.with(
+        OpenTelemetry::Trace::Tracer.const_get(:CONTEXT_SPAN_KEY),
+        current_span
+      ) do
```
I feel like this is testing at the wrong level of abstraction. The `OpenTelemetry::Context` is really an internal concern, along with the `Tracer::CONTEXT_SPAN_KEY`. It should be possible to test the semantics of `tracer.current_span` using only the public API of `Tracer`. In particular, the `in_span` convenience method, `with_span` and `current_span` are all related, and I think it makes sense to test them as a unit rather than using gymnastics to test them separately.

elskwid replied:
@fbogsany it's a clunky test for sure. Let me poke at it a bit and see if I can make it better. Thanks for the guidance.

elskwid replied:
@fbogsany I've settled on using `with_span` to test `current_span` so I moved the `current_span` expectation up to this `describe` block and removed it from the `#with_span` block. I think that makes it clearer what I'm testing.

fbogsany commented on api/test/opentelemetry/trace/tracer_test.rb:
```diff
@@ -0,0 +1,118 @@
+      end
+
+      finished.must_equal(true)
+    end
```
I think you could test `in_span` by subclassing `Tracer`, overriding `start_span` to return a mock span, and set expectations on it. Some of the techniques in `batch_span_processor_test.rb` and `simple_span_processor_test.rb` seem useful here.

elskwid replied:
@fbogsany I sure can! I'll see what I can do. Thanks.

fbogsany commented:
> I like that the Python API makes the call to finish/end optional. If we're interested in that I can open a new PR for that functionality. @fbogsany mentioned we may want to do this.

I'm going to walk back my 'like' for this. This is a convenience API that now wraps 3 things that a client can already do with the API:
1. start a new span
1. make it active
1. finish it
Optionally removing one of the 3 doesn't really bring enough value to justify the complication to the method signature. Compare:
```ruby
span = nil
result = OpenTelemetry.tracer.in_span('do-the-thing', finish: false) do |new_span|
  span = new_span
  compute_the_thing
end
...
span.finish
```
with:
```ruby
span = OpenTelemetry.tracer.start_span('do-the-thing')
result = OpenTelemetry.tracer.with_span(span) { compute_the_thing }
...
span.finish
```
I'd argue the latter is cleaner, easier to understand, and easier to use.

fbogsany commented on api/lib/opentelemetry/trace/tracer.rb:
```diff
@@ -30,6 +30,7 @@ def current_span
       def in_span(name, attributes: nil, links: nil, start_timestamp: nil, kind: nil, sampling_hint: nil)
         span = start_span(name, attributes: attributes, links: links, start_timestamp: start_timestamp, kind: kind, sampling_hint: sampling_hint)
         with_span(span) { |s| yield s }
+        span.finish
```
This needs to capture the result of the `with_span` (technically, the result of its block) and return it after the `span.finish`. `.tap` may be appropriate here.

elskwid replied:
@fbogsany I think I see a way to do that. Thanks for the behavior check!

fbogsany replied:
It just occurred to me that this should be in an `ensure`. I.e.:
```ruby
  with_span(span) { |s| yield s }
ensure
  span.finish
end
```
which has the added advantage of not requiring an extra local variable.

elskwid commented:
> > I like that the Python API makes the call to finish/end optional. If we're interested in that I can open a new PR for that functionality. @fbogsany mentioned we may want to do this.
> 
> I'm going to walk back my 'like' for this. This is a convenience API that now wraps 3 things that a client can already do with the API:
> 
>     1. start a new span
> 
>     2. make it active
> 
>     3. finish it
>        Optionally removing one of the 3 doesn't really bring enough value to justify the complication to the method signature. Compare:
> 
> 
> ```ruby
> span = nil
> result = OpenTelemetry.tracer.in_span('do-the-thing', finish: false) do |new_span|
>   span = new_span
>   compute_the_thing
> end
> ...
> span.finish
> ```
> 
> with:
> 
> ```ruby
> span = OpenTelemetry.tracer.start_span('do-the-thing')
> result = OpenTelemetry.tracer.with_span(span) { compute_the_thing }
> ...
> span.finish
> ```
> 
> I'd argue the latter is cleaner, easier to understand, and easier to use.

@fbogsany Totally agree with this. Looking closer, the Python folks have added the optional flag to `Tracer.use_span` which appears to map to our `Tracer#with_span`. It wasn't clear to me that the method would accept a `Span`. (The [API uses `span: "Span"`](https://github.com/open-telemetry/opentelemetry-python/blob/master/opentelemetry-api/src/opentelemetry/trace/__init__.py#L531) while the [SDK uses `span: trace_api.Span`](https://github.com/open-telemetry/opentelemetry-python/blob/master/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L382)). Maybe this changes our opinion of the usefulness...

Keep `#in_span` signature the way it is. 

```ruby
OpenTelemetry.tracer.in_span('do-the-thing') do |new_span|
  compute_the_thing
end
```

`#with_span` using default `finish` of `false`:

```ruby
span = OpenTelemetry.tracer.start_span('do-the-thing')
result = OpenTelemetry.tracer.with_span(span) { compute_the_thing }
# ...
span.finish
```

`#with_span` using `finish ==  true`:

```ruby
span = OpenTelemetry.tracer.start_span('do-the-thing')
result = OpenTelemetry.tracer.with_span(span, finish: true) { compute_the_thing }
```

Since `#in_span` uses `#with_span` it can default to `finish: true`, and that might also fix up the problem of returning the value from the `with_span` block.

fbogsany commented:
I think it is worth laying out the common use-cases of each method:
* `start_root_span` is used when explicitly starting a new trace. This is not common for typical HTTP services, but can be used in a background (e.g. Resque or Sidekiq) job, and may also be used for services that form a boundary between 3rd-party and 1st-party systems (where we may not want to continue a trace started by another business entity). The common pattern is something like:
```ruby
def foo
  span = tracer.start_root_span('foo')
  ... # maybe tracer.with_span(span) at some point
ensure
  span.finish
end
```
* `start_span` provides the most flexibility. It is typically used when a caller wishes to precisely manage the lifetime of the span, the span parent, and also wishes to manage its "activation" via `with_span` (or not, as the case may be). Typical patterns include:
```ruby
def foo(parent_span)
  span = tracer.start_span('foo', with_parent: parent_span)
  ... # maybe tracer.with_span(span) at some point
ensure
  span.finish
end

def foo
  span = tracer.start_span('foo') # implicit parent span from current context
  Thread.new(span) do |span|
    ... # maybe tracer.with_span(span) at some point
  ensure
    span.finish
  end
end
```
* `with_span` simply activates and deactivates the span w.r.t. the current context. The primary value here is making the "current span" available through a call chain that does not propagate the span via an explicit parameter. Examples include pretty much any Ruby framework at present, which obviously do not anticipate the existence of OpenTelemetry. Example:
```ruby
def leaf_method
  span = start_span('leaf') # implicit parent span from current context
  do_the_thing
ensure
  span.finish
end

def framework_method
  leaf_method
end

def service_method
  wire_context = tracer.http_text_format.extract(headers) { |headers, key| headers[key] } # what a horrible API, in retrospect
  span = tracer.start_span('service', with_parent_context: wire_context)
  result = tracer.with_span(span) { framework_method }
  format_result(result)
ensure
  span.finish
end
```
* `in_span` is a convenience method for a common use-case, especially in manual instrumentation (the moral equivalent of `StatsD.measure('my-metric') { ... }` or manual timing code & logging). The most common use-case outside of auto-instrumentation is wrapping a block of code in a span, where the author shouldn't have to think about propagation (where her span's parent comes from, which spans her span parents). E.g.:
```ruby
def my_wonderful_but_maybe_slow_method
  tracer.in_span('my_wonderful_but_maybe_slow_method') do
    sleep(30)
    do_something
  end
end
```
* `current_span` is most often used when a new piece of information is discovered at a random point in the code and we want to add it as an attribute or event to the current span (effectively add a metric/dimension or a log), without really caring about what that span is. E.g.
```ruby
def i_just_figured_it_out!
  tracer.current_span['answer'] = 42
end
```

From this perspective, I don't think `with_span` should do optional-double-duty of finishing the span. Its sole purpose is to propagate the span across API boundaries ("in process context propagation"). Most leaf methods (e.g. when instrumenting calls to Redis, memcached, MySQL, etc.) should use explicit span creation and `finish`.

elskwid commented:
@fbogsany I believe I have addressed all your feedback. This is rebased on master and ready to go. Thank you again for the thorough review and all the help. I've opened issue #123 to update the documentation with your `Tracer` method explanations above.

🙇

fbogsany commented on api/test/opentelemetry/trace/tracer_test.rb:
```diff
@@ -0,0 +1,124 @@
+        span.must_equal(wrapper_span)
+      end
+    end
+  end
```
Both `with_span` and `in_span` should exhibit the following behaviour:
```ruby
span_a = tracer.start_span('a')
span_b = tracer.start_span('b')
tracer.with_span(span_a) do
  tracer.current_span.must_equal(span_a)
  tracer.with_span(span_b) do
    tracer.current_span.must_equal(span_b)
  end
  tracer.current_span.must_equal(span_a)
end
```
i.e. the span that was active before the block is re-activated after the block. We don't have tests for that here.

elskwid replied:
@fbogsany great example! I'll get that in.

fbogsany commented on api/test/opentelemetry/trace/tracer_test.rb:
```diff
@@ -0,0 +1,124 @@
+  describe '#start_root_span' do
+    it 'requires a name' do
+      proc { tracer.start_root_span(nil) }.must_raise(ArgumentError)
+    end
```
This will go away with #115

fbogsany commented on api/test/opentelemetry/trace/tracer_test.rb:
```diff
@@ -0,0 +1,124 @@
+
+    it 'requires a name' do
+      proc { tracer.start_span(nil) }.must_raise(ArgumentError)
+    end
```
This will go away with #115

fbogsany commented on api/test/opentelemetry/trace/tracer_test.rb:
```diff
@@ -0,0 +1,124 @@
+
+    it 'returns a span with the same context as the parent' do
+      span = tracer.start_span('op', with_parent: parent)
+      span.context.must_equal(parent.context)
```
This behaviour is wrong. Or maybe it isn't. It's definitely up for discussion, with the Named Tracers change. The SDK does not behave in this way, and I don't think the API's "no-op" implementation should either. Specifically, the new Span should share the parent span's trace ID, but should have a different span ID.

https://github.com/open-telemetry/opentelemetry-ruby/blob/3b52ae0beae634e6ab18db3b50c1bb4825b98c8f/api/lib/opentelemetry/trace/tracer.rb#L48-L50

vs.

https://github.com/open-telemetry/opentelemetry-ruby/blob/3b52ae0beae634e6ab18db3b50c1bb4825b98c8f/sdk/lib/opentelemetry/sdk/trace/tracer.rb#L102

fbogsany replied:
Let's stick with the current behaviour for now. This matches the Java implementation of [`DefaultSpan`](https://github.com/open-telemetry/opentelemetry-java/blob/master/api/src/main/java/io/opentelemetry/trace/DefaultSpan.java). There's some related debate in https://github.com/open-telemetry/opentelemetry-specification/issues/208

fbogsany commented on api/test/opentelemetry/trace/tracer_test.rb:
```diff
@@ -0,0 +1,124 @@
+    it 'returns an instance of HTTPTextFormat' do
+      tracer.http_text_format.must_be_instance_of(Propagation::HTTPTextFormat)
+    end
+  end
```
These will move to `tracer_format_test.rb` once #122 lands.

elskwid replied:
I've moved it.Thanks!

mwear commented:
It looks like #122 has landed, and this is conflicted. What do we need to do to move this forward?

elskwid commented:
@mwear I will make the moves on Friday and see if I can get it over the line.

elskwid commented:
@fbogsany I've addressed all the suggestions/issues you raised with the exception of changing the behavior you mentioned in [this comment](https://github.com/open-telemetry/opentelemetry-ruby/pull/118#discussion_r335766500).

@mwear this has been rebased and the Minitest deprecations have also been addressed.

I believe it's good to go.
